### ISSUE TITLE
Incorrect MSE reported on status bar when l2_lambda is not zero.

### ISSUE THREAD

**Issue by sergeyf** [fchollet/keras#464]
Consider the following toy example dataset and network:

```
from __future__ import print_function
import numpy as np
from keras.models import Graph
from keras.layers.core import Dense
from keras.regularizers import l2

# generate random data
d = 6000
X1 = np.random.random((10000,d))**2
X2 = np.log(np.random.random((10000,d)))
Y = (np.dot(X1,np.random.random((d,1))) - np.dot(X2,np.random.random((d,1))))**2
Y /= Y.max() # scale to be between 0 and 1

data = {'X1':X1, 'X2':X2, 'output':Y}

# network parameters
d1 = 512
d2 = 256
l2_lambda = 1e-3

# graph model
model = Graph()

# inputs
model.add_input(name='X1', ndim=2)
model.add_input(name='X2', ndim=2)

# X1 dense layer
model.add_node(Dense(d, d1, activation='relu',W_regularizer=l2(l2_lambda)), 
               name='dense_X1', input='X1')

# X2 dense layer
model.add_node(Dense(d, d1, activation='relu',W_regularizer=l2(l2_lambda)), 
               name='dense_X2', input='X2')

# merging dense layer
model.add_node(Dense(2*d1 , d2, activation='relu',W_regularizer=l2(l2_lambda)), 
               name='dense_merge', merge_mode="concat", 
               inputs=['dense_X1','dense_X2'])

# output dense layer
model.add_node(Dense(d2 , 1, activation='sigmoid',W_regularizer=l2(l2_lambda)), 
               name='dense_final',input='dense_merge')      
               
model.add_output(name='output', input="dense_final")

model.compile('rmsprop', {"output": 'mse'})
```

First, I check the MSE of the network BEFORE it is trained. 

```
predictions = model.predict(data)
print('MSE before any training:')
print(np.mean((predictions['output']-Y)**2))

MSE before any training:
0.444015524597
```

The MSE before training is 0.44. So, once we actually start training, we would expect the progress bar to report something in that vicinity.

However, during the actual training the ```output``` is nonsensically huge:

```
history = model.fit(data=data, nb_epoch=3, validation_split=0.25)
Train on 7500 samples, validate on 2500 samples
Epoch 0
7500/7500 [==============================] - 2s - output: 2.1041 - val_output: 0.0096
Epoch 1
7500/7500 [==============================] - 2s - output: 1.6632 - val_output: 0.0096
```

Note the `output: 2.1041` which is MUCH bigger than the val_output. How can the MSE be this large?

I have noticed that this bug is not active if we change `l2_lambda` to be 0:

```
MSE before any training:
0.0487085829439

Train on 7500 samples, validate on 2500 samples
Epoch 0
7500/7500 [==============================] - 2s - output: 0.0093 - val_output: 0.0083
Epoch 1
7500/7500 [==============================] - 2s - output: 0.0084 - val_output: 0.0083
```

Any idea what's going on here?

**Comment by iskandr** [comment]
Utterly mystifying that a regularization parameter should cause the MSE to get rescaled to values greater than 1.

**Comment by sergeyf** [comment]
Also, if you just have a single dense layer, the problem is much less evident:

```
from __future__ import print_function
import numpy as np
from keras.models import Graph
from keras.layers.core import Dense
from keras.regularizers import l2

# generate random data
d = 6000
X1 = np.random.random((10000,d))**2
X2 = np.log(np.random.random((10000,d)))
Y = (np.dot(X1,np.random.random((d,1))) - np.dot(X2,np.random.random((d,1))))**2
Y /= Y.max()

data = {'X':np.concatenate((X1,X2),axis=1), 'output':Y}

# network parameters
d1 = 512
l2_lambda = 1e-3

# graph model
model = Graph()

# inputs
model.add_input(name='X', ndim=2)

# X1 dense layer
model.add_node(Dense(2*d, 1, activation='sigmoid',W_regularizer=l2(l2_lambda)), 
               name='dense_X', input='X')
     
model.add_output(name='output', input="dense_X")

model.compile('rmsprop', {"output": 'mse'})
```

And then MSE before and during training is sensible:

```
MSE before any training:
0.525251362627
Train on 7500 samples, validate on 2500 samples
Epoch 0
7500/7500 [==============================] - 1s - output: 0.0190 - val_output: 0.0078
Epoch 1
7500/7500 [==============================] - 1s - output: 0.0097 - val_output: 0.0078
```

So now `output` is lower than MSE before training, but still twice as large as `val_output`, which generally doesn't happen.

**Comment by iskandr** [comment]
@fchollet Any idea why this happens?

**Comment by fchollet** [comment]
> Utterly mystifying that a regularization parameter should cause the MSE to get rescaled to values greater than 1.

Utterly normal. Regularization happens by incorporating the regularization term into the loss. The score reported is no longer just MSE. The loss value you're seeing is MSE + the regularization parameter. 

Hence it's higher than the MSE. If it's *significantly* higher than MSE, that means you need to reduce the L2 factor in order to bring it back to a reasonable range, which is necessary for learning to happen smoothly.

**Comment by fchollet** [comment]
> So now output is lower than MSE before training, but still twice as large as val_output, which generally doesn't happen.

Loss scores from random inputs through a random network tend to be random, unsurprisingly. Try it with a new random seed and you'll see completely different results.

**Comment by sergeyf** [comment]
So the what's being reported on the training data is the entire regularized loss function, but what about on the validation data, which was much smaller in the original example? I assume it is smaller because the model.predict is being used. If so, these two outputs are inconsistent (one is the whole regularized loss and one is the raw MSE), and will probably confuse many people. I think it is generally expected that the training loss and the validation loss have are computed with the same function. 

Perhaps it would be good to add a flag so that the kind of loss being shown for training and validation has a similar interpretation.

**Comment by fchollet** [comment]
> So the what's being reported on the training data is the entire regularized loss function, but what about on the validation data, which was much smaller in the original example? 

Regularization is only incorporated in the train loss function, since it is only meant to help training. The validation loss is still MSE. The network itself also behaves differently during training and validation (e.g. dropout turned off), so it's very common to see different scores in training and validation.

**Comment by sergeyf** [comment]
I see the logic, even though I disagree with the choice of what to display on the screen. The issue is that the current set-up is unintuitive given that the two differently computed outputs are reported adjacent to each other using the same vocabulary without clarifying further text. Here is what I (or any user) sees:

```output: 2.1041 - val_output: 0.0096```

My suggestion is to do something like this

```training_loss: 2.1041 - training_error: 0.0095 - val_error: 0.0096```

**Comment by RyanCarey** [comment]
Yes, by default the training error (unregularized) and validation error should be displayed (and the regularized training error too, though it's least important) so that you can see whether you're overfitting - the most important aspect of NN monitoring.
